Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp

```diff
@@ -0,0 +1,185 @@
+{
+    if (m_instance != XR_NULL_HANDLE)
+        xrDestroyInstance(m_instance);
+}
+
+void OpenXRCoordinator::getPrimaryDeviceInfo(WebPageProxy&, DeviceInfoCallback&& callback)
+{
+    ASSERT(RunLoop::isMain());
+
+    initializeDevice();
+    if (m_instance == XR_NULL_HANDLE || m_systemId == XR_NULL_SYSTEM_ID) {
+        LOG(XR, "Failed to initialize OpenXR system");
+        callback(std::nullopt);
+        return;
+    }
+
+    auto supportsOrientationTracking = [instance = m_instance, system = m_systemId]() -> bool {
+        XrSystemProperties systemProperties = createOpenXRStruct<XrSystemProperties, XR_TYPE_SYSTEM_PROPERTIES>();
+        CHECK_XRCMD(xrGetSystemProperties(instance, system, &systemProperties));
+        return systemProperties.trackingProperties.orientationTracking == XR_TRUE;
+    };
+
+    XRDeviceInfo deviceInfo { .identifier = m_deviceIdentifier, .vrFeatures = { }, .arFeatures = { } };
+    deviceInfo.supportsOrientationTracking = supportsOrientationTracking();
+    deviceInfo.supportsStereoRendering = m_currentViewConfiguration == XR_VIEW_CONFIGURATION_TYPE_PRIMARY_STEREO;
+    deviceInfo.recommendedResolution = recommendedResolution();
+    LOG(XR, "OpenXR device info:\n\tOrientation tracking: %s\n\tStereo rendering: %s\n\tRecommended resolution: %dx%d", deviceInfo.supportsOrientationTracking ? "yes" : "no", deviceInfo.supportsStereoRendering ? "yes" : "no", deviceInfo.recommendedResolution.width(), deviceInfo.recommendedResolution.height());
+
+    // OpenXR runtimes MUST support VIEW and LOCAL reference spaces.
+    deviceInfo.vrFeatures.append(PlatformXR::SessionFeature::ReferenceSpaceTypeViewer);
+    deviceInfo.arFeatures.append(PlatformXR::SessionFeature::ReferenceSpaceTypeViewer);
+    deviceInfo.vrFeatures.append(PlatformXR::SessionFeature::ReferenceSpaceTypeLocal);
+    deviceInfo.arFeatures.append(PlatformXR::SessionFeature::ReferenceSpaceTypeLocal);
+
+    if (m_extensions->isExtensionSupported(XR_MSFT_UNBOUNDED_REFERENCE_SPACE_EXTENSION_NAME ""_span)) {
+        deviceInfo.vrFeatures.append(PlatformXR::SessionFeature::ReferenceSpaceTypeUnbounded);
+        deviceInfo.arFeatures.append(PlatformXR::SessionFeature::ReferenceSpaceTypeUnbounded);
+    }
+
+    // In order to get the supported reference space types, we need the session to be created. However at this point we shouldn't do it.
+    // Instead, we report ReferenceSpaceTypeLocalFloor as available, because we can supoport it via either the STAGE reference space, the
+    // LOCAL_FLOOR reference space or even via an educated guess from the LOCAL reference space as a backup.
+    deviceInfo.vrFeatures.append(PlatformXR::SessionFeature::ReferenceSpaceTypeLocalFloor);
+    deviceInfo.arFeatures.append(PlatformXR::SessionFeature::ReferenceSpaceTypeLocalFloor);
+
+    callback(WTFMove(deviceInfo));
+}
+
+void OpenXRCoordinator::createInstance()
+{
+    ASSERT(RunLoop::isMain());
+    ASSERT(m_instance == XR_NULL_HANDLE);
+
+    Vector<char *> extensions;
```

Comment:
```suggestion
    Vector<const char*, 2> extensions;
```

Reply:
OK, but we'll add more extensions in future patches